### ISSUE TITLE
elixir's setupHook causes compiler to emit warnings for all code

### ISSUE THREAD

**Issue by Sorixelle** [NixOS/nixpkgs#85265]
**Describe the bug**
The `setupHook` in `elixir` is causing all compilation of Elixir source files to emit warnings for files in Elixir's standard library. This only occurs when the compiler is being run inside of `nix-shell`, and does not occur when `elixir` is installed globally, eg. through `environment.systemPackages` in NixOS.

**To Reproduce**
1. Start a shell containing Elixir with `nix-shell -p elixir`
2. Run `mix new test` to generate a sample Elixir project
3. cd into the new `test` directory and run `mix compile` to compile the project.
4. Observe the large amount of warnings emitted, as shown [here](https://pastebin.com/w6gXj3XJ).

**Expected behavior**
The compiler should not emit the warnings it does, and simply output only the following:
```
Compiling 1 file (.ex)
Generated test app
```

**Additional context**
This issue seems to be caused by setting `ERL_LIBS` in Elixir's `setupHook`. Unsetting this environment variable while in the `nix-shell` does cause the compiler warnings to go away. I'm not sure what the purpose of setting this variable is, so I don't know if the solution is as simple as just removing the `setupHook`, or if there's a bit more nuance here.

This issue affects Elixir 1.8 - 1.10. Elixir 1.6 and 1.7 are not affected by this. Elixir 1.8 in particular emits absolute paths beginning with `/build/source`, which looks to me like a remnant of the build environment.

**Notify maintainers**
@the-kenny @Havvy @couchemar @ankhers @filalex77 

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.6.2, NixOS, 20.09pre219796.f601ab37c2f (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.3`
 - channels(root): `"home-manager, nixos-20.09pre219796.f601ab37c2f"`
 - channels(ruby): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
  - elixir
  - elixir_1_10
  - elixir_1_9
  - elixir_1_8
# a list of nixos modules affected by the problem
module:
```

**Comment by cw789** [comment]
This issue was already mentioned here:

- [discourse.nixos.org: Elixir: compilation warnings only in nix-shell – ERL_LIBS](https://discourse.nixos.org/t/elixir-compilation-warnings-only-in-nix-shell-erl-libs/3917)
- [elixirforum.com: Compilation warnings (clause cannot match) in mix-and-otp tutorial](https://elixirforum.com/t/compilation-warnings-clause-cannot-match-in-mix-and-otp-tutorial/25114)
- [discourse.nixos.org: State of the BEAM Ecosystem in Nix](https://discourse.nixos.org/t/state-of-the-beam-ecosystem-in-nix/4202)

I personally for now set the needed ERL_LIBS manually for my environments:
```
with import <nixpkgs> {};
let
  elixir = pkgs.buildPackages.beam.packages.erlangR22.elixir_1_10;
  hex = pkgs.buildPackages.beam.packages.erlangR22.hex;
in
pkgs.mkShell {
  LANG = "C.UTF8";
  CLEANED_ERL_LIBS = "${hex}/lib/erlang/lib";
  buildInputs = [elixir hex];
  shellHook = ''
    export ERL_LIBS=$CLEANED_ERL_LIBS
  '';
}
```

So I'm not sure if setting ERL_LIBS is still really needed for Elixir.

@DianaOlympos
Did you find out something new about this topic?

**Comment by Sorixelle** [comment]
I think ERL_LIBS is used to facilitate building Elixir apps/libraries through Nix, since (if I'm understanding the BEAM ecosystem in Nix correctly) each Mix dependency is intended to be built as an individual derivation, and provided through ERL_LIBS to work around Mix's network-dependent dependency fetching/resolving. Although, I don't see any scenario where having Elixir itself in ERL_LIBS would be useful.

**Comment by DianaOlympos** [comment]
no i have not dived into this one yet and tbh i will not have time until at best end of May.

Thanks for the exploration though. I would love for someone to try to understand that part though. Most of the stuff happening in there is still quite opaque to me tbh.

**Comment by happysalada** [comment]
In my latest test, I have not observed the issue. If this is still an issue, please let me know and I can have a look

**Comment by cw789** [comment]
@happysalada Hi. I can still reproduce the issue described above.
`ERL_LIBS` is still set within the Elixir generic-builder causing the long duplicated list of compiler warnings.

https://github.com/NixOS/nixpkgs/blob/d10236eb20163b9a5871a45bada37886b2bc1d39/pkgs/development/interpreters/elixir/generic-builder.nix#L29

https://github.com/NixOS/nixpkgs/blob/d10236eb20163b9a5871a45bada37886b2bc1d39/pkgs/development/interpreters/elixir/setup-hook.sh#L1-L5

This means the following lands in `ERL_LIBS`:
```
ls -l /nix/store/4plz7d3nfl8hmlm568cgf91l0a4wj828-elixir-1.11.4/lib/elixir/lib

dr-xr-xr-x 3 main main 4.0K Jan  1  1970 eex
dr-xr-xr-x 3 main main 4.0K Jan  1  1970 elixir
dr-xr-xr-x 3 main main 4.0K Jan  1  1970 ex_unit
dr-xr-xr-x 3 main main 4.0K Jan  1  1970 iex
dr-xr-xr-x 3 main main 4.0K Jan  1  1970 logger
dr-xr-xr-x 3 main main 4.0K Jan  1  1970 mix
```

But if I `unset ERL_LIBS` I can still use e.g. `iex` within my nix-shell.
So this is still not clear to me, why this was added to the Elixir build script by @ericbmerritt?
Commit: https://github.com/NixOS/nixpkgs/commit/2d6d9682bb8a910f3eb5f0be8d5687c753e9732c

---

As I tried now to dig in deeper, a new guess of mine is, it will somehow affect the usage of escripts.
If you have a look at the mix escript.build - Configuration - **`:embed_elixir`**:
→ [hexdocs.pm/mix/Mix.Tasks.Escript.Build.html#module-configuration](https://hexdocs.pm/mix/Mix.Tasks.Escript.Build.html#module-configuration)

As we see `:embed_elixir` was from the beginning of Elixir v1.0 always set default to `true`,
it would probably not affect many if we remove setting `ERL_LIBS` within the Elixir generic-builder?

But as I still do not understand the real reason why it was added - I might be totally wrong!
